airgun/widgets.py

```diff
@@ -916,31 +916,44 @@ class CustomParameter(Table):
 
     add_new_value = Text("..//a[contains(text(),'+ Add Parameter')]")
 
-    def __init__(self, parent, locator=None, id=None, logger=None):
+    # def __init__(self, parent, locator=None, id=None, logger=None, column_widgets=None):
+    def __init__(self, parent, **kwargs):
         """Supports initialization via ``locator=`` or ``id=``"""
-        if locator and id or not locator and not id:
+        if (
+            kwargs.get('locator')
+            and kwargs.get('id')
+            or not kwargs.get('locator')
+            and not kwargs.get('id')
+        ):
             raise ValueError('Please specify either locator or id')
-        locator = locator or f".//table[@id='{id}']"
+        locator = kwargs.get('locator') or f".//table[@id='{kwargs.get('id')}']"
 
-        column_widgets = {
+        column_widgets = kwargs.get('column_widgets') or {
             'Name': TextInput(locator=".//input[@placeholder='Name']"),
             'Value': TextInput(locator=".//textarea[@placeholder='Value']"),
             'Actions': Text(
                 locator=".//a[@data-original-title='Remove Parameter' "
                 "or @title='Remove Parameter']"
             ),
         }
-        super().__init__(parent, locator=locator, logger=logger, column_widgets=column_widgets)
+        self.name = list(column_widgets.keys())[0]
+        self.name_key = '_'.join(self.name.lower().split(' '))
+        self.value = list(column_widgets.keys())[1]
+        self.value_key = '_'.join(self.value.lower().split(' '))
+
+        super().__init__(
```

Comment:
Good work handling the kwargs above!

This call to super() isn't quite right though - no need to try and match up kwargs with the parent class, just pass the whole thing.

```
super().__init__(parent, **kwargs)
```

Reply:
@mshriver So I've tried changing the super() call to `super().__init__(parent, **kwargs) both in its existing location and with it moved to the first line in the this `__init__` method, but both are introducing errors in some of the other tests that use the `CustomParameter` widget. As an example, `tests/foreman/ui/test_domain.py::test_positive_set_parameter` fails with `TypeError: Table.__init__() got an unexpected keyword argument 'id'`. The same test passes with the current super() call. I'm wondering if a change to the `Table` widget would be necessary for this to work, but I'm a bit leery of making a change to such a widely-used widget. What do you think is the best approach here?

Reply:
So I'm still having issues with this. Making the changes that we discussed has my Ansible tests still passing, but it's failing on tests that use views that instantiate the `CustomParameter` class with an `id` argument (these are the UI test modules `test_operatingsystem.py`, `test_host.py`, and `test_domain.py`). The issue seems to be that `locator` in the table widget's `__init__` method is a positional argument with no default value. However, passing a locator argument to `super().__init__` does not resolve the issue. Instead, the Ansible test now fails, and the tests that use `CustomParameter` with `id` still fail, now because the widget seems to be evaluating to a `NoneType` object. Any ideas on how to work past this?

Here are two of the variations that I tried along with links to the tracebacks that I encountered:

```
    def __init__(self, parent, **kwargs):
        """Supports initialization via ``locator=`` or ``id=``"""
        if (
            kwargs.get('locator')
            and kwargs.get('id')
            or not kwargs.get('locator')
            and not kwargs.get('id')
        ):
            raise ValueError('Please specify either locator or id')
        locator = kwargs.get('locator') or f".//table[@id='{kwargs.pop('id')}']"
        super().__init__(parent, **kwargs)
```
results:
ansible test passes
error in host test: http://rx-paste.usersys.redhat.com/view/834c2c5f


```
    def __init__(self, parent, **kwargs):
        """Supports initialization via ``locator=`` or ``id=``"""
        if (
            kwargs.get('locator')
            and kwargs.get('id')
            or not kwargs.get('locator')
            and not kwargs.get('id')
        ):
            raise ValueError('Please specify either locator or id')
        locator = kwargs.get('locator') or f".//table[@id='{kwargs.pop('id')}']"
        super().__init__(parent, locator=locator, **kwargs)

```
results:
error in ansible test: http://rx-paste.usersys.redhat.com/view/2842dfc7
error in host test: http://rx-paste.usersys.redhat.com/view/ee7b053b